server/src/Utils/Order.hs

```diff
@@ -0,0 +1,38 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Utils.Order where
+
+import Import.NoFoundation
+import Model.Types
+
+-- {-| Determine if a given user is allowed to access the PDF file of an item.
+-- -}
+hasAccessToPdfFileDownload ::
+       ( BaseBackend (YesodPersistBackend site) ~ SqlBackend
+       , YesodPersist site
+       , PersistQueryRead (YesodPersistBackend site)
+       )
+    => UserId
+    -> Text
+    -> HandlerT site IO AuthResult
+hasAccessToPdfFileDownload userId filename = do
+    let unauthorized = Unauthorized "You didn't buy this item"
```

Comment:
@zarvunk I'd appreciate your help with this (non-compiling) as-well. The idea of the function is to do a bunch of queries, where each query depends on the success of the previous one.

I could write it in a very verbose manner with a bunch of nested `case .. of`, but I wonder how it would be best written.